edge-util/src/Microsoft.Azure.Devices.Edge.Util/ClientTokenProvider.cs

```diff
@@ -38,11 +43,33 @@ public class ClientTokenProvider : ITokenProvider
 
         public async Task<string> GetTokenAsync(Option<TimeSpan> ttl)
         {
-            DateTime startTime = DateTime.UtcNow;
+            return await this.tokenCache.GetToken().Match(
+                t => Task.FromResult(t),
+                async () =>
+                {
+                    using (await this.cacheUpdateLock.LockAsync())
```

Comment:
Ok, from what I see you are locking here to avoid locking every single call

Reply:
Correct.